ws-discovery/networking.go

```diff
@@ -40,46 +41,63 @@ func GetAvailableDevicesAtSpecificEthernetInterface(interfaceName string) []onvi
 	if strings.TrimSpace(interfaceName) == "" {
 		interfaceName = defaultInterfaceName
 	}
-	devices := SendProbe(interfaceName, nil, nil, map[string]string{"dn": "http://www.onvif.org/ver10/network/wsdl"})
+	probeResponses := SendProbe(interfaceName, nil, nil, map[string]string{"dn": "http://www.onvif.org/ver10/network/wsdl"})
 	nvtDevices := make([]onvif.Device, 0)
+	nvtDevices, err := devicesFromProbeResponses(probeResponses)
+	if err != nil {
+		fmt.Printf("Fail to discover Onvif camera: %s \n", err)
+	}
+
+	return nvtDevices
+}
 
-	for _, j := range devices {
+func devicesFromProbeResponses(probeResponses []string) ([]onvif.Device, error) {
+	nvtDevices := make([]onvif.Device, 0)
+	xaddrSet := make(map[string]struct{})
+	for _, j := range probeResponses {
 		doc := etree.NewDocument()
 		if err := doc.ReadFromString(j); err != nil {
-			fmt.Errorf("%s", err.Error())
-			return nil
+			return nil, fmt.Errorf("%s", err.Error())
 		}
 
-		endpoints := doc.Root().FindElements("./Body/ProbeMatches/ProbeMatch/XAddrs")
-		for _, xaddr := range endpoints {
-			xaddr := strings.Split(strings.Split(xaddr.Text(), " ")[0], "/")[2]
-			fmt.Printf("Onvif WS-Discovery: Find %s \n", xaddr)
-			c := 0
-
-			for c = 0; c < len(nvtDevices); c++ {
-				if nvtDevices[c].GetDeviceParams().Xaddr == xaddr {
-					fmt.Println(nvtDevices[c].GetDeviceParams().Xaddr, "==", xaddr)
-					break
+		probeMatches := doc.Root().FindElements("./Body/ProbeMatches/ProbeMatch")
+		for _, probeMatch := range probeMatches {
+			var xaddr string
+			if address := probeMatch.FindElement("./XAddrs"); address != nil {
+				u, err := url.Parse(address.Text())
+				if err != nil {
+					fmt.Printf("Invalid XAddrs: %s\n", address.Text())
+					continue
 				}
+				xaddr = u.Host
 			}
-
-			if c < len(nvtDevices) {
+			if _, dupe := xaddrSet[xaddr]; dupe {
+				fmt.Printf("Skipping duplicate XAddr: %s\n", xaddr)
 				continue
 			}
 
-			dev, err := onvif.NewDevice(onvif.DeviceParams{Xaddr: strings.Split(xaddr, " ")[0]})
+			var endpointRefAddress string
+			if ref := probeMatch.FindElement("./EndpointReference/Address"); ref != nil {
+				uuidElements := strings.Split(ref.Text(), ":")
+				uuid := uuidElements[len(uuidElements)-1]
+				endpointRefAddress = uuid
```

Comment:
superfluous variable
```suggestion
				endpointRefAddress = uuidElements[len(uuidElements)-1]
```

Reply:
Modified.